File: app/src/main/java/com/islam/basepropject/project_base/base/other/BaseViewModel.java
package com.islam.basepropject.project_base.base.other;

import androidx.lifecycle.ViewModel;

import com.islam.basepropject.project_base.base.POJO.ErrorModel;
import com.islam.basepropject.project_base.base.POJO.ScreenStatus;
import com.islam.basepropject.project_base.utils.others.SchedulerProvider;

import java.lang.ref.WeakReference;
import java.util.HashMap;

import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Consumer;
import io.reactivex.subjects.PublishSubject;

public abstract class BaseViewModel<N> extends ViewModel {

    private final SchedulerProvider mSchedulerProvider;

    private final CompositeDisposable mCompositeDisposable;
    private final PublishSubject<String> mSnackBarMessage;
    private final PublishSubject<String> mToastMessage;
    private final PublishSubject<String> mDialogMessage;
    private final PublishSubject<Boolean> mShowLoadingFullScreen;
    private final PublishSubject<ErrorModel> mShowErrorFullScreen;
    private final HashMap<String, ScreenStatus> registeredFragments;
    private WeakReference<N> mNavigator;
    private String lastRegisteredFragment;

    public BaseViewModel() {
        this.mSchedulerProvider = new SchedulerProvider();
        this.mCompositeDisposable = new CompositeDisposable();
        mSnackBarMessage = PublishSubject.create();
        mToastMessage = PublishSubject.create();
        mDialogMessage = PublishSubject.create();
        mShowLoadingFullScreen = PublishSubject.create();
        mShowErrorFullScreen = PublishSubject.create();
        registeredFragments = new HashMap<>();
    }

    public void registerFragment(String fragmentClassName) {
        if (!registeredFragments.containsKey(fragmentClassName))
            registeredFragments.put(fragmentClassName, ScreenStatus.STARTING);
        lastRegisteredFragment = fragmentClassName;
    }

    public ScreenStatus getLastRegisterdFragmentStatus() {
        return registeredFragments.get(lastRegisteredFragment);
    }


    @Override
    protected void onCleared() {
        mCompositeDisposable.dispose();
        super.onCleared();
    }

    protected void addDisposable(Disposable disposable) {
        mCompositeDisposable.add(disposable);
    }


    public N getNavigator() {
        return mNavigator.get();
    }

    public void setNavigator(N navigator) {
        this.mNavigator = new WeakReference<>(navigator);
    }

    public SchedulerProvider getSchedulerProvider() {
        return mSchedulerProvider;
    }

    public void showSnackBarMessage(String s) {
        mSnackBarMessage.onNext(s);
    }

    public void showToastMessage(String s) {
        mToastMessage.onNext(s);
    }

    public void showDialogMessage(String s) {
        mDialogMessage.onNext(s);
    }

    public void showLoadingFullScreen(boolean b) {
        mShowLoadingFullScreen.onNext(b);
    }

    public void showNoConnectionFullScreen(ErrorModel errorModel) {
        mShowErrorFullScreen.onNext(errorModel);
    }

    //to prevent dublicate observing because observeDefault called twice in OnViewCreated and in onStart
    public boolean isDefaultObserved() {
        return mDialogMessage.hasObservers()
                && mToastMessage.hasObservers()
                && mSnackBarMessage.hasObservers();
    }


    public void observeDialogMessage(Consumer<String> consumer) {
        addDisposable(mDialogMessage.subscribe(consumer));
    }

    public void observeToastMessage(Consumer<String> consumer) {
        addDisposable(mToastMessage.subscribe(consumer));
    }

    public void observeSnackBarMessage(Consumer<String> consumer) {
        addDisposable(mSnackBarMessage.subscribe(consumer));
    }

    public void observeShowLoadingFullScreen(Consumer<Boolean> consumer) {
        addDisposable(mShowLoadingFullScreen.subscribe(consumer));

    }

    public void observeShowNoConnectionFullScreen(Consumer<ErrorModel> consumer) {
        addDisposable(mShowErrorFullScreen.subscribe(consumer));
    }

    public void unSubscribe() {
        mCompositeDisposable.clear();
    }

    public void unRegister(String fragmentClassName) {
        registeredFragments.remove(fragmentClassName);
    }
}

File: app/src/main/java/com/islam/basepropject/project_base/base/other/BaseViewModel.kt
package com.islam.basepropject.project_base.base.other

//import androidx.lifecycle.ProcessLifecycleOwner
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ProcessLifecycleOwner
import androidx.lifecycle.ViewModel
import androidx.lifecycle.lifecycleScope
import com.islam.basepropject.MyApplication
import com.islam.basepropject.project_base.POJO.ErrorModel
import com.islam.basepropject.project_base.POJO.Message
import com.islam.basepropject.project_base.POJO.ScreenStatus
import com.islam.basepropject.project_base.base.other.network.Failure
import com.islam.basepropject.project_base.base.other.network.Result
import com.islam.basepropject.project_base.base.other.network.RetrofitCorounites
import com.islam.basepropject.project_base.utils.navigation.Navigator
import java.util.*
import kotlin.collections.List
import kotlin.collections.MutableMap
import kotlin.collections.get
import kotlin.collections.mutableMapOf
import kotlin.collections.set
import kotlin.collections.toMutableMap

abstract class BaseViewModel : ViewModel() {

    /*
     Do not forget to check if liveData value is null before requesting the data
     Use appScope when you need the response of network request even that user has left the screen
     Use viewModelScope when the network response don't represent any importance if the user left the screen
    */

    val appScope = ProcessLifecycleOwner.get().lifecycleScope
    val mLoadingViews = MutableLiveData<MutableMap<Int, Boolean>>()//list of view that may show loading instead show full screen loading
    val mSnackBarMessage = SingleLiveEvent<Message>()
    val mEnableSensitiveInputs = MutableLiveData<Boolean>()
    val mToastMessage = SingleLiveEvent<Message>()
    val mDialogMessage = SingleLiveEvent<Message>()
    val mShowLoadingFullScreen = MutableLiveData<Boolean>()
    val mShowErrorFullScreen = MutableLiveData<ErrorModel>()
    private val registeredFragments = HashMap<String, ScreenStatus>()
    private var lastRegisteredFragment: String? = null

    val lastRegisterdFragmentStatus: ScreenStatus?
        get() = registeredFragments[lastRegisteredFragment]

    init {
        this.loadInitialData()
        mLoadingViews.value = mutableMapOf()
    }

    open fun loadInitialData() {}

    fun registerFragment(fragmentClassName: String) {

        if (!registeredFragments.containsKey(fragmentClassName))
            registeredFragments[fragmentClassName] = ScreenStatus.STARTING
        lastRegisteredFragment = fragmentClassName
    }

    //if all network jobs finished successfully mark the screen as completed
    fun markAsCompleted(results: List<Result<Any>>) {
        for (result in results) {
            if (result is Failure)
                return
        }
        lastRegisteredFragment?.let { registeredFragments[it] = ScreenStatus.COMPLETED }

    }


    suspend fun <T> networkCall(viewId: Int? = null, block: suspend () -> T): Result<T> {
        return RetrofitCorounites(this, viewId).networkCall(block)
    }


    fun markAsCompleted(name: String) {
        registeredFragments[name] = ScreenStatus.COMPLETED
    }


    fun showSnackBarMessage(s: Message) {
        mSnackBarMessage.value = s
    }

    fun showToastMessage(s: Message) {
        if (mToastMessage.hasActiveObservers())
            mToastMessage.value = s
        else
            showAlertBar(s)
    }

    fun showDialogMessage(s: Message) {
        mDialogMessage.value = s
    }

    fun enableSensitiveInputs(b: Boolean) {
        mEnableSensitiveInputs.value = b
    }

    fun showLoadingFullScreen(b: Boolean) {
        mShowLoadingFullScreen.value = b
    }

    fun showNoConnectionFullScreen(errorModel: ErrorModel?) {
        mShowErrorFullScreen.value = errorModel
    }

    fun showAlertBar(message: Message) {
        MyApplication.instance?.showAlertBar(message)
    }

    private val isStartingFragment: Boolean
        get() = lastRegisterdFragmentStatus == ScreenStatus.STARTING

    fun showLoading(viewId: Int?) {
        enableSensitiveInputs(false)
        if (isStartingFragment) {
            showNoConnectionFullScreen(null)
            showLoadingFullScreen(true)
        } else if (viewId != null) {
            mLoadingViews.value?.put(viewId, true)
            mLoadingViews.value = mLoadingViews.value?.toMutableMap()
        }
    }

    fun hideLoading(viewId: Int?) {
        enableSensitiveInputs(true)
        if (isStartingFragment)
            showLoadingFullScreen(false)
        else if (viewId != null) {
            mLoadingViews.value?.put(viewId, false)
            mLoadingViews.value = mLoadingViews.value
        }
    }


    fun unRegister(fragmentClassName: String) {
        //registeredFragments.remove(fragmentClassName)
        if (fragmentClassName == lastRegisteredFragment)
            lastRegisteredFragment = null // MAY CAUSE PROBLEM
    }

}
